sycl/doc/design/spirv-extensions/SPV_INTEL_fpmath.asciidoc

```diff
@@ -0,0 +1,129 @@
+SPV_INTEL_fpmath
+================
+
+== Name Strings
+
+SPV_INTEL_fpmath
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/KhronosGroup/SPIRV-Registry
+
+== Contributors
+
+- Shuo Niu, Intel
+
+== Notice
+
+Copyright (c) 2022 Intel Corporation.  All rights reserved.
+
+== Status
+
+Final draft
+
+== Version
+
+[width="40%",cols="25,25"]
+|========================================
+| Last Modified Date | 2022-02-23
+| Revision           | 1
+|========================================
+
+== Dependencies
+
+This extension is written against the SPIR-V Specification,
+Version 1.6 Revision 1.
+
+This extension requires SPIR-V 1.0.
+
+== Overview
+
+This extension adds a decoration that may be attached to any instruction of floating-point type. It can be used to express the maximum acceptable error in the result of that instruction, in ULPs.
```

Comment:
Do we need to describe how the decoration interacts with the minimum ULP requirements described in the various environment specifications?  Presumably the decoration "wins"?

Reply:
Had an offline discussion with @GarveyJoe about this. Perhaps the decoration should always win in SPIR-V so that it can express both less or more accuracy. But this would make it illegal to drop the decoration in SPIR-V even though it is legal to drop it in LLVM IR.

Add @andykaylor for more input.

Reply:
That's a really good question and it has implications for where the decoration comes from in LLVM IR. I have been working on a proposal to clarify the semantics of the LLVM IR "fpmath" metadata, and as you say because it is metadata it has to be legal to drop it. The default semantics of a math library function call in LLVM IR are implementation-defined accuracy. So, for OpenCL and SYCL targets the accuracy required by the environment is the default accuracy without the "fpmath" metadata and is what you will get if the metadata is dropped. Therefore the definition of the LLVM IR metadata has to say that the metadata cannot require greater accuracy than the implementation-defined default.

For SYCL, however, I expect increased accuracy to be an important use can, though one that not every device will be able to support.

It seems to me that we really need two different things here: (1) permitted error, and (2) required accuracy. To support required accuracy, we'd need a way to query the device for available accuracy and fail if the required accuracy is not available. Permitted error is important for increasing the limits beyond the normal specification so that we can take advantage of faster hardware capabilities, but I know some people want the ability to restrict the maximum error more than the SYCL specification does.

For LLVM IR, I think the required accuracy has to be an attribute so that it can't be dropped. For SPIRV, I think we could define it so a single decorator handles both required accuracy and permitted error, but we won't be able to use LLVM's "fpmath" metadata for that.

Reply:
> For SPIRV, I think we could define it so a single decorator handles both required accuracy and permitted error, but we won't be able to use LLVM's "fpmath" metadata for that.

Do you still expect to keep this fpmath equivalent SPIR-V decoration then?

Reply:
@shuoniu-intel I don't have a proposal for the LLVM IR replacement. That will require some discussion to reach a consensus. I expect that we will end up wanting the SPIR-V decorator as you have proposed here, but we might want to hold off on committing it until we see what happens with the IR.

Reply:
Got it.